### PR TITLE
[LIR][SCEVExpander] Restore original flags when aborting transform

### LINKED ISSUE #82337: [LoopIdiomRecognize] Poison flags stripped even though no transform performed
In this example, LIR will remove the nuw/exact flags on `%dist` and `%len`, even though it does not actually perform any transform:
```llvm
; RUN: opt -S -passes=loop-idiom < %s
define void @test(ptr %p.end, ptr %p.start) {
entry:
  %p.end.int = ptrtoint ptr %p.end to i64
  %p.start.int = ptrtoint ptr %p.start to i64
  %dist = sub nuw i64 %p.end.int, %p.start.int
  %len = lshr exact i64 %dist, 5
  %cmp = icmp eq ptr %p.end, %p.start
  br i1 %cmp, label %exit, label %preheader

preheader:
  br label %loop

loop:
  %iv = phi i64 [ %len, %preheader ], [ %iv.dec, %loop ]
  %iv.neg = sub nsw i64 0, %iv
  %src = getelementptr inbounds [32 x i8], ptr %p.end, i64 %iv.neg
  %v = load <4 x i64>, ptr %src, align 8
  %dst = getelementptr inbounds [32 x i8], ptr %p.start, i64 %iv.neg
  store <4 x i64> %v, ptr %dst, align 8
  %iv.dec = add i64 %iv, -1
  %cmp2 = icmp eq i64 %iv.dec, 0
  br i1 %cmp2, label %loop.exit, label %loop

loop.exit:
  br label %exit

exit:
  ret void
}
```

### PR DESCRIPTION
**Pull request by nikic**
SCEVExpanderCleaner will currently remove instructions created by SCEVExpander, but not restore poison generating flags that it may have dropped. As such, running LIR can currently spuriously drop flags without performing any transforms.

Fix this by keeping track of original instruction flags in SCEVExpander.

Fixes https://github.com/llvm/llvm-project/issues/82337.

### CODE REVIEW

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

File: llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp [hunk 1]
```diff
@@ -43,6 +43,37 @@ cl::opt<unsigned> llvm::SCEVCheapExpansionBudget(
...
 
+PoisonFlags::PoisonFlags(const Instruction *I) {
+  NUW = false;
+  NSW = false;
+  Exact = false;
+  Disjoint = false;
+  NNeg = false;
+  if (auto *OBO = dyn_cast<OverflowingBinaryOperator>(I)) {
+    NUW = OBO->hasNoUnsignedWrap();
+    NSW = OBO->hasNoUnsignedWrap();
```
Comment (antoniofrighetto) [inline llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:54]: Shouldn't this be `hasNoSignedWrap`? Same as below when setting.

Comment (nikic) [reply llvm/lib/Transforms/Utils/ScalarEvolutionExpander.cpp:54]: Indeed! Thanks for catching it.

**Comment by nikic** [comment]
@nathanchance Thanks for the report! Here's a reduced test case:
```llvm
; RUN: opt -S -passes=loop-reduce < %s
target triple = "riscv64-unknown-linux-gnu"

define void @test(ptr %p, i8 %arg, i32 %start) {
entry:
  %conv = zext i8 %arg to i32
  %shr = lshr i32 %conv, 1
  %wide.trip.count = zext nneg i32 %shr to i64
  br label %for.body

for.body:
  %indvars.iv = phi i64 [ 0, %entry ], [ %indvars.iv.next, %for.body ]
  %add810 = phi i32 [ %start, %entry ], [ %add, %for.body ]
  %idxprom2 = zext i32 %add810 to i64
  %arrayidx3 = getelementptr i8, ptr %p, i64 %idxprom2
  %v = load i8, ptr %arrayidx3, align 1
  %add = add i32 %add810, 1
  %indvars.iv.next = add i64 %indvars.iv, 1
  %exitcond.not = icmp eq i64 %indvars.iv, %wide.trip.count
  br i1 %exitcond.not, label %exit, label %for.body

exit:
  ret void
}
```

**Comment by nikic** [comment]
The issue should be fixed by https://github.com/llvm/llvm-project/commit/72763521c34287bce68402eb2a9d71dcb4eed5a0.

**Pull request merged**